### PR TITLE
Add RotateAbsolute action to turtlesim

### LINKED ISSUE #61: [ROS 2] Add actions to turtlesim
Turtlesim doesn't have any actions, but I think it's important for the sake of the tutorials that it include at least one simple action. Actions are one of the core concepts that need to be introduced, and it would damage the tutorial set if we had to switch to another example just for the sake of explaining actions. 

One simple action with some kind of feedback that users can see is all that's needed.

### PR DESCRIPTION
**Pull request by jacobperron**
Resolves #61

### CODE REVIEW

**Comment by dirk-thomas** [comment]
In order to trigger this action easily it would be good to update keyboard teleop with a few key to rotate to e.g. orientations in 90 (e.g. R|G|V|D) / 45 (e.g. R|T|G|B|V|C|D|E) degrees steps.

**Comment by jacobperron** [comment]
> In order to trigger this action easily it would be good to update keyboard teleop with a few key to rotate to e.g. orientations in 90 (e.g. R|G|V|D) / 45 (e.g. R|T|G|B|V|C|D|E) degrees steps.

I was just about to post some commands for testing this. Since this is intended to go along with the tutorial, I figured the user will be prompted to send the actions with `ros2 action` tool. For example:

    ros2 action send_goal /turtle1/rotate_absolute turtlesim/action/RotateAbsolute "theta: 1.57"

another with feedback:

    ros2 action send_goal /turtle1/rotate_absolute turtlesim/action/RotateAbsolute "theta: 1.0" --feedback

Adding to the keyboard teleop isn't a bad idea either. Want me to follow up in a second PR?

**Comment by dirk-thomas** [comment]
> Want me to follow up in a second PR?

Either a separate one or added to this. I don't mind either way.

**Comment by jacobperron** [comment]
@dirk-thomas See 0e6dec2 for keyboard teleop support. I also made the 'F' key cancel an action.

File: turtlesim/src/turtle.cpp [hunk 1]
```diff
@@ -34,13 +34,20 @@
 #include <QColor>
 #include <QRgb>
 
+#define TWO_PI 6.283185307179586
```
Comment (dirk-thomas) [inline turtlesim/src/turtle.cpp]: Why not use the definition of `PI` from `turtlesim/turtle.h`?

**Comment by dirk-thomas** [comment]
With this patch when quiting the teleop node the console is "unusable" since the state isn't correctly restored anymore. E.g. you don't see what you type. Since that is not the case on the target branch it seems to be related to the changes.

**Comment by jacobperron** [comment]
> With this patch when quiting the teleop node the console is "unusable" since the state isn't correctly restored anymore. E.g. you don't see what you type. Since that is not the case on the target branch it seems to be related to the changes.

Odd. It doesn't happen for me in ZSH, but I can reproduce the issue with Bash...

**Comment by jacobperron** [comment]
@dirk-thomas  c31fab3 seems to resolve the shell issue after shutdown.

File: turtlesim/tutorials/teleop_turtle_key.cpp [hunk 2]
```diff
@@ -145,20 +243,30 @@ int main(int argc, char** argv)
 
   int rc = teleop_turtle.keyLoop();
   input.shutdown();
-  rclcpp::shutdown();
```
Comment (dirk-thomas) [inline turtlesim/tutorials/teleop_turtle_key.cpp]: Why is this being removed?

Comment (jacobperron) [reply turtlesim/tutorials/teleop_turtle_key.cpp]: It shouldn't have been, my mistake. Restored 55c5aa4

File: turtlesim/src/turtle.cpp (hunk 1 above)
Comment (jacobperron) [reply turtlesim/src/turtle.cpp]: See 5f614f0

**Comment by dirk-thomas** [comment]
Thanks for the new functionality. Looks good to me :+1:

File: turtlesim/tutorials/teleop_turtle_key.cpp [hunk 3]
```diff
@@ -55,29 +67,74 @@ class KeyboardReader
...
         }
         else if (buffer.Event.KeyEvent.wVirtualKeyCode == VK_RIGHT)
         {
-          *c = KEYCODE_R;
+          *c = KEYCODE_RIGHT;
           return;
         }
         else if (buffer.Event.KeyEvent.wVirtualKeyCode == VK_DOWN)
+        {
+          *c = KEYCODE_DDOWN;
```
Comment (dirk-thomas) [inline turtlesim/tutorials/teleop_turtle_key.cpp:85]: `undeclared identifier`: see #66 for a fix.

File: turtlesim/tutorials/teleop_turtle_key.cpp [hunk 4]
```diff
@@ -123,6 +184,43 @@ TeleopTurtle::TeleopTurtle():
   nh_->get_parameter("scale_linear", l_scale_);
 
   twist_pub_ = nh_->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 1);
+  rotate_absolute_client_ = rclcpp_action::create_client<turtlesim::action::RotateAbsolute>(nh_, "turtle1/rotate_absolute");
+}
+
+void TeleopTurtle::sendGoal(double theta)
+{
+  auto goal = turtlesim::action::RotateAbsolute::Goal();
+  goal.theta = theta;
```
Comment (dirk-thomas) [inline turtlesim/tutorials/teleop_turtle_key.cpp:193]: @jacobperron https://ci.ros2.org/job/ci_windows/8159/warnings43Result/new/

> '=': conversion from 'double' to 'turtlesim::action::RotateAbsolute_Goal_>::_theta_type', possible loss of data

Comment (jacobperron) [reply turtlesim/tutorials/teleop_turtle_key.cpp:193]: See https://github.com/ros/ros_tutorials/pull/69 for a fix

**Pull request merged**